tensorflow/core/kernels/sparse_tensor_dense_matmul_op.cc

```diff
@@ -262,53 +298,68 @@ struct SparseTensorDenseMatMulFunctor<CPUDevice, T, Tindices, ADJ_A, ADJ_B> {
+        Eigen::array<Eigen::Index, NDIM> indices;
+        for (int j = 0; j < NDIM - 2; ++j) {
+          indices[j] = a_indices(i, j);
+        }
         for (std::size_t n = 0; n < rhs_right; ++n) {
-          const T b_value = maybe_adjoint_b(k, n);
-          out(m, n) += a_value * b_value;
+          indices[NDIM - 2] = k;
+          indices[NDIM - 1] = n;
+          const T b_value = maybe_adjoint_b(indices);
+          indices[NDIM - 2] = m;
+          out(indices) += a_value * b_value;
         }
       }
     } else {
       // Vectorization via Eigen.
-      const int b_chip_index = ADJ_B ? 1 : 0;
-
-#define LOOP_NNZ(b_passed)                                                  \
-  for (std::size_t i = 0; i < nnz; ++i) {                                   \
-    const Tindices m = internal::SubtleMustCopy(a_indices(i, lhs_index_a)); \
-    const Tindices k = internal::SubtleMustCopy(a_indices(i, rhs_index_a)); \
-    const T a_value = (ADJ_A) ? MaybeConj(a_values(i)) : a_values(i);       \
-    if (!FastBoundsCheck(k, lhs_right)) {                                   \
-      return KOutOfBoundsError(k, i, rhs_index_a, lhs_right);               \
-    }                                                                       \
-    if (!FastBoundsCheck(m, out.dimension(0))) {                            \
-      return MOutOfBoundsError(m, i, lhs_index_a, out.dimension(0));        \
-    }                                                                       \
-    out.template chip<0>(m) +=                                              \
-        b_passed.template chip<b_chip_index>(k) * a_value;                  \
-  }
-
-      if (ADJ_B) {
-        // Perform transpose and conjugation on B once, since we chip out B's
-        // columns in the nnz loop.
-        Eigen::array<int, 2> shuffle(1, 0);  // preserve dimension order
-        Eigen::Tensor<T, 2, Eigen::ColMajor> col_major_conj_b =
-            b.swap_layout().shuffle(shuffle).conjugate();
-        LOOP_NNZ(col_major_conj_b);
-      } else {
-        LOOP_NNZ(b);
+      for (std::size_t i = 0; i < nnz; ++i) {
+        const Tindices m = internal::SubtleMustCopy(a_indices(i, lhs_index_a));
+        const Tindices k = internal::SubtleMustCopy(a_indices(i, rhs_index_a));
+        const T a_value = (ADJ_A) ? MaybeConj(a_values(i)) : a_values(i);
+        if (!FastBoundsCheck(k, lhs_right)) {
+          return KOutOfBoundsError(k, i, rhs_index_a, lhs_right);
+        }
+        if (!FastBoundsCheck(m, out.dimension(NDIM - 2))) {
+          return MOutOfBoundsError(m, i, lhs_index_a, out.dimension(NDIM - 2));
+        }
+        Eigen::array<Eigen::Index, NDIM> out_offsets;
+        Eigen::array<Eigen::Index, NDIM> out_extends;
+        Eigen::array<Eigen::Index, NDIM> b_offsets;
+        Eigen::array<Eigen::Index, NDIM> b_extends;
+        for (int j = 0; j < NDIM - 2; ++j) {
+          out_offsets[j] = a_indices(i, j);
```

Comment:
As above, since you have to use SubtleMustCopy for security reasons, perhaps it is easier to copy all the indices at the top of the i-loop?